### PR TITLE
expression: add simple expression semantic equal check logic

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/43572

Problem Summary:

### What is changed and how it works?
In TiDB Expression tree, sometimes we may need judge whether two expression tree are semantic equal,
this PR adds the simple check logic ported from SparkSQL.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: expression/scalar_function.go [hunk 1]
```diff
@@ -438,13 +439,119 @@ func (sf *ScalarFunction) EvalJSON(ctx sessionctx.Context, row chunk.Row) (types
...
 	}
 	ReHashCode(sf, sc)
 	return sf.hashcode
 }
 
+func ExpressionsSemanticEqual(ctx sessionctx.Context, expr1, expr2 Expression) bool {
+	sc := ctx.GetSessionVars().StmtCtx
+	sc.CanonicalHashCode.Store(true)
+	defer sc.CanonicalHashCode.Store(false)
+	return string(expr1.HashCode(sc)) == string(expr2.HashCode(sc))
```
Comment (qw4990) [inline expression/scalar_function.go:457]: Using `bytes.Equal(...)` seems better. `string(...)` can incur some unnecessary data copy and object allocation.

File: expression/scalar_function.go [hunk 2]
```diff
@@ -438,13 +439,119 @@ func (sf *ScalarFunction) EvalJSON(ctx sessionctx.Context, row chunk.Row) (types
...
+		return sf.hashcode
+	}
 	if len(sf.hashcode) > 0 {
 		return sf.hashcode
 	}
 	ReHashCode(sf, sc)
 	return sf.hashcode
 }
 
+func ExpressionsSemanticEqual(ctx sessionctx.Context, expr1, expr2 Expression) bool {
```
Comment (qw4990) [inline expression/scalar_function.go:453]: Better to add some comments for this exposed function to make CI happy...

**Review by qw4990 — APPROVED** [review]

File: expression/scalar_function.go [hunk 3]
```diff
@@ -438,13 +439,120 @@ func (sf *ScalarFunction) EvalJSON(ctx sessionctx.Context, row chunk.Row) (types
...
 	ReHashCode(sf, sc)
 	return sf.hashcode
 }
 
+// ExpressionsSemanticEqual is used to judge whether two expression tree is semantic equivalent.
+func ExpressionsSemanticEqual(ctx sessionctx.Context, expr1, expr2 Expression) bool {
+	sc := ctx.GetSessionVars().StmtCtx
+	sc.CanonicalHashCode.Store(true)
+	defer sc.CanonicalHashCode.Store(false)
+	return bytes.Equal(expr1.HashCode(sc), expr2.HashCode(sc))
```
Comment (xuyifangreeneyes) [inline expression/scalar_function.go:458]: `(*ScalarFunction).HashCode` is not a pure function and has the side effect(i.e., if `len(sf.hashcode) > 0`, then we don't need to recalculate it). Is it ok to expose the hash code calculated under `CanonicalHashCode` to outside?

File: expression/scalar_function.go [hunk 4]
```diff
@@ -438,13 +439,120 @@ func (sf *ScalarFunction) EvalJSON(ctx sessionctx.Context, row chunk.Row) (types
...
+			for i := len(argsHashCode) - 1; i >= 0; i-- {
+				sf.hashcode = append(sf.hashcode, argsHashCode[i]...)
+			}
+		}
+	case ast.UnaryNot:
+		child, ok := sf.GetArgs()[0].(*ScalarFunction)
+		if !ok {
+			// use the origin arg hash code.
+			for _, argCode := range argsHashCode {
+				sf.hashcode = append(sf.hashcode, argCode...)
```
Comment (xuyifangreeneyes) [inline expression/scalar_function.go:511]: Do we need to encode the not op?

Comment (AilinKid) [reply expression/scalar_function.go:511]: Nope：
Expression1: not(a<b) 
Expression2: a>=b 
Expression1 and Expression2 should have the same hashcode, we try to disintegrate the NOT into a mathematical operation convertion.

File: expression/scalar_function.go (hunk 3 above)
Comment (AilinKid) [reply expression/scalar_function.go:462]: Nice catch!
Addressed.

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**